asyncio_run_in_process/_child.py

```diff
@@ -70,11 +72,69 @@ def update_state_finished(to_parent: BinaryIO, finished_payload: bytes) -> None:
     to_parent.flush()
 
 
+async def _do_task_cleanup(*tasks: 'asyncio.Future[Any]') -> None:
+    for task in tasks:
+        if not task.done():
+            task.cancel()
+
+        try:
+            await task
+        except asyncio.CancelledError:
+            pass
+
+
 SHUTDOWN_SIGNALS = {signal.SIGTERM}
 
 
+async def _handle_SIGTERM(task: 'asyncio.Task[TReturn]', fut: 'asyncio.Future[int]') -> None:
+    signum = await fut
+    raise SystemExit(signum)
+
+
+async def _handle_coro(coro: Coroutine[Any, Any, TReturn], got_SIGINT: asyncio.Event) -> TReturn:
```

Comment:
There's a lot of asyncio knowledge I had to swap back in, in order to understand what this method was doing. It might be worth adding some of them as comments:

1. `asyncio.wait()` wraps all co-routines passed to it in `asyncio.Task`s. `asyncio.Task`s _cannot* be abandoned, they must be awaited. This method calls `_do_task_cleanup` in order to clean up the created Tasks before the method exits.
2. coroutines are "just" generators, calling `coro.throw(exc)` resumes execution of the coroutine by raising the given exception. (From inside the coro it looks as if the `await` call raised the exception)
3. Another behavior which comes from generators: if `coro.throw()` raises `StopIteration`, it means that the coroutine finished running without hitting any other `await` calls. This is equivalent to the coroutine finishing, and it can no longer be awaited or cancelled or a `RuntimeError` will be thrown.

Reply:
Yeah, this was a particularly complex/hard thing to figure out.

I did try to leave a number of detailed comments through the body of this function since it's doing some very nuanced and low level things.  I'm not sure if it makes sense to try and explain the nature of coroutines here as well because this code isn't going to be understandable without that base knowledge and I don't see this as the right place to try and include that.

I'm not firmly against it so I'll play with adding something in a docstring that tries to capture the overall gist of this function.

Reply:
see c17837f